src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs

```diff
@@ -804,68 +805,86 @@ private void PinSocketAddressBuffer()
         }
 
         // Cleans up any existing Overlapped object and related state variables.
-        private void FreeOverlapped(bool checkForShutdown)
+        private void FreeOverlapped()
         {
-            if (!checkForShutdown || !Environment.HasShutdownStarted)
+            // Free the overlapped object.
+            if (_ptrNativeOverlapped != null && !_ptrNativeOverlapped.IsInvalid)
```

Comment:
I think we'd ideally dispose of it even if it is invalid.  SafeHandles will still be finalized if they haven't been disposed, even if they contain the invalid value.